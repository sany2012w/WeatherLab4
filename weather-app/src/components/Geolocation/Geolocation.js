import React from 'react';
import './Geolocation.css';
import WeatherData from '../WeatherData/WeatherData'
import { connect } from "react-redux";
import { setCoords, fetchGeolocationError, setLoadingTrue } from '../../actions/geolocationAction';
import { fetchWeatherByCoords } from '../../actions/fetchWeatherByCoords';
import Header from '../Header/Header'
import Loader from '../Loader/Loader'

class Geolocation extends React.Component {
  componentDidMount() {
    this.getGeolocation();
  }

  render() {
    let weatherBlock = this.setWeatherBlock();
    return (
      <div>
        <Header updateGeolocation = {this.getGeolocation.bind(this)}/>
        {!this.props.isLoading && weatherBlock}
        {this.props.isLoading && <Loader/>}
      </div>
    );
  }

  getGeolocation() {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(position => {
        this.props.setLoadingTrue();
        this.props.setCoords({lat: position.coords.latitude, lon: position.coords.longitude});
        this.props.fetchWeatherByCoords(this.props.coords);
      },
      () => {
        this.props.setCoords({lat: 51.51, lon: -0.13});
        this.props.fetchWeatherByCoords(this.props.coords);
        alert('Permission denied. Load weather from default coordinates');
      });
    } else {
      this.props.fetchGeolocationError('your browser does not support geolocation');
    }
  }

  setWeatherBlock ()  {
    let weatherBlock = null;
    if(this.props.forecast) {
      weatherBlock = <WeatherData data={this.props.forecast}/>;
    } else if (this.props.error){
      weatherBlock = <div className="error">Error: {this.props.error}</div>;
    } else if (!this.props.forecast && !this.props.error) {
      weatherBlock = <Loader/>
    }
    return weatherBlock;
  }

}


function mapStateToProps(state) {
return {
  coords: state.geolocation.coords,
  forecast: state.geolocation.forecast,
  error: state.geolocation.error,
  isLoading: state.geolocation.isLoading
};
}

function mapDispatchToProps(dispatch) {
return {
  setCoords: (coords) => {
    dispatch(setCoords(coords));
  },

  fetchWeatherByCoords: (coords) => {
    dispatch(fetchWeatherByCoords(coords));
  },

  setLoadingTrue: () => {
    dispatch(setLoadingTrue());
  },

  fetchGeolocationError: (error) => {
    dispatch(fetchGeolocationError(error));
  }
};
}

export default connect(mapStateToProps, mapDispatchToProps)(Geolocation);
