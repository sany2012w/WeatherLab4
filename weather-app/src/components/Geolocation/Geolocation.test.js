import React from "react";
import Geolocation from "./Geolocation";
import renderer from "react-test-renderer";
import weatherTestData from "./weatherTestData";
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux'
import thunk from "redux-thunk";

const testStore = configureMockStore([thunk]);

it("Weather data loaded", () => {
  const store = testStore({
    geolocation: {
          forecast: weatherTestData,
          isLoading: false
      }
    });

  const tree = renderer.create(
    <Provider store={store}>
      <Geolocation/>
   </Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Weather data is loading and no forecast", () => {
  const store = testStore({
    geolocation: {
          isLoading: true
      }
    });

  const tree = renderer.create(
    <Provider store={store}>
      <Geolocation/>
   </Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Weather data is loading and there is forecast", () => {
  const store = testStore({
    geolocation: {
          forecast: weatherTestData,
          isLoading: true
      }
    });

  const tree = renderer.create(
    <Provider store={store}>
      <Geolocation/>
   </Provider>).toJSON();

  expect(tree).toMatchSnapshot();
});

it("Error", () => {
  const store = testStore({
    geolocation: {
          error: 'Something wrong'
        }
    });

  const tree = renderer.create(
    <Provider store={store}>
      <Geolocation/>
   </Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("There are only coords without forecast", () => {
  const store = testStore({
    geolocation: {
          coords: {
            lat: 10,
            lon: 51
          }
        }
    });

  const tree = renderer.create(
    <Provider store={store}>
      <Geolocation/>
   </Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("There are coords with forecast", () => {
  const store = testStore({
    geolocation: {
          coords: {
            lat: 10,
            lon: 51
          },
          forecast: weatherTestData
        }
    });

  const tree = renderer.create(
    <Provider store={store}>
      <Geolocation/>
   </Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});
