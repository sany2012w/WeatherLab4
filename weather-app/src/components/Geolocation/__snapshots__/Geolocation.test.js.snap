// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error 1`] = `
<div>
  <div
    className="header"
  >
    <div
      className="headerText"
    >
      Weather here
    </div>
    <button
      className="headerButton"
      onClick={[Function]}
    >
      Update Geolocation
    </button>
  </div>
  <div
    className="error"
  >
    Error: 
    Something wrong
  </div>
</div>
`;

exports[`There are coords with forecast 1`] = `
<div>
  <div
    className="header"
  >
    <div
      className="headerText"
    >
      Weather here
    </div>
    <button
      className="headerButton"
      onClick={[Function]}
    >
      Update Geolocation
    </button>
  </div>
  <div
    className="WeatherBlock"
  >
    <div
      className="City"
    >
      London
    </div>
    <div
      className="leftWeather"
    >
      <img
        className="Icon"
        src="https://openweathermap.org/img/wn/09d.png"
      />
      <div
        className="Temp"
      >
        7
         °C
      </div>
    </div>
    <div
      className="rightWeather"
    >
      <div
        className="Pressure"
      >
        Pressure: 
        1012
         hPa
      </div>
      <div
        className="Humidity"
      >
        Humidity: 
        81
        %
      </div>
      <div
        className="Clouds"
      >
        Clouds: 
        light intensity drizzle
      </div>
      <div
        className="Wind"
      >
        Wind: 
        4.1
         m/s
      </div>
      <div
        className="Coords"
      >
        Coordinates: [
        -0.13
        , 
        51.51
        ]
      </div>
    </div>
  </div>
</div>
`;

exports[`There are only coords without forecast 1`] = `
<div>
  <div
    className="header"
  >
    <div
      className="headerText"
    >
      Weather here
    </div>
    <button
      className="headerButton"
      onClick={[Function]}
    >
      Update Geolocation
    </button>
  </div>
  <div
    className="wrapper"
  >
    <div
      className="loader-circle"
    />
    <div
      className="loader-bottom"
    >
      Please wait, weather is loading...
    </div>
  </div>
</div>
`;

exports[`Weather data is loading and no forecast 1`] = `
<div>
  <div
    className="header"
  >
    <div
      className="headerText"
    >
      Weather here
    </div>
    <button
      className="headerButton"
      onClick={[Function]}
    >
      Update Geolocation
    </button>
  </div>
  <div
    className="wrapper"
  >
    <div
      className="loader-circle"
    />
    <div
      className="loader-bottom"
    >
      Please wait, weather is loading...
    </div>
  </div>
</div>
`;

exports[`Weather data is loading and there is forecast 1`] = `
<div>
  <div
    className="header"
  >
    <div
      className="headerText"
    >
      Weather here
    </div>
    <button
      className="headerButton"
      onClick={[Function]}
    >
      Update Geolocation
    </button>
  </div>
  <div
    className="wrapper"
  >
    <div
      className="loader-circle"
    />
    <div
      className="loader-bottom"
    >
      Please wait, weather is loading...
    </div>
  </div>
</div>
`;

exports[`Weather data loaded 1`] = `
<div>
  <div
    className="header"
  >
    <div
      className="headerText"
    >
      Weather here
    </div>
    <button
      className="headerButton"
      onClick={[Function]}
    >
      Update Geolocation
    </button>
  </div>
  <div
    className="WeatherBlock"
  >
    <div
      className="City"
    >
      London
    </div>
    <div
      className="leftWeather"
    >
      <img
        className="Icon"
        src="https://openweathermap.org/img/wn/09d.png"
      />
      <div
        className="Temp"
      >
        7
         °C
      </div>
    </div>
    <div
      className="rightWeather"
    >
      <div
        className="Pressure"
      >
        Pressure: 
        1012
         hPa
      </div>
      <div
        className="Humidity"
      >
        Humidity: 
        81
        %
      </div>
      <div
        className="Clouds"
      >
        Clouds: 
        light intensity drizzle
      </div>
      <div
        className="Wind"
      >
        Wind: 
        4.1
         m/s
      </div>
      <div
        className="Coords"
      >
        Coordinates: [
        -0.13
        , 
        51.51
        ]
      </div>
    </div>
  </div>
</div>
`;
